nomad/server.go

```diff
@@ -353,6 +355,92 @@ func NewServer(config *Config, consulCatalog consul.CatalogAPI, logger *log.Logg
 	return s, nil
 }
 
+func getTLSConf(enableRPC bool, tlsConf *tlsutil.Config) (*tls.Config, tlsutil.RegionWrapper, error) {
+	var tlsWrap tlsutil.RegionWrapper
+	var incomingTLS *tls.Config
+	if enableRPC {
+		tw, err := tlsConf.OutgoingTLSWrapper()
+		if err != nil {
+			return nil, nil, err
+		}
+		tlsWrap = tw
+
+		itls, err := tlsConf.IncomingTLSConfig()
+		if err != nil {
+			return nil, nil, err
+		}
+		incomingTLS = itls
+	}
+	return incomingTLS, tlsWrap, nil
+}
+
+// ReloadTLSConnections updates a server's TLS configuration and reloads RPC
+// connections
+func (s *Server) ReloadTLSConnections(newTLSConfig *config.TLSConfig) error {
+	s.logger.Printf("[INFO] nomad: reloading server connections due to configuration changes")
+
+	s.configLock.Lock()
+	s.config.TLSConfig = newTLSConfig
+	s.configLock.Unlock()
+
+	tlsConf := s.config.tlsConfig()
+	incomingTLS, tlsWrap, err := getTLSConf(s.config.TLSConfig.EnableRPC, tlsConf)
+	if err != nil {
+		s.logger.Printf("[ERR] nomad: unable to reset TLS context")
+		return err
+	}
+
+	if s.rpcCancel == nil {
+		s.logger.Printf("[ERR] nomad: No TLS Context to reset")
+		return fmt.Errorf("Unable to reset tls context")
+	}
+
+	s.rpcCancel()
+
+	s.rpcTLSLock.Lock()
+	s.rpcTLS = incomingTLS
+	s.rpcTLSLock.Unlock()
+
+	s.raftTransportLock.Lock()
+	defer s.raftTransportLock.Unlock()
+	s.raftTransport.Close()
+
+	s.connPool.ReloadTLS(tlsWrap)
+
+	// reinitialize our rpc listener
+	s.rpcListenerLock.Lock()
+	s.rpcListener.Close()
+	time.Sleep(500 * time.Millisecond)
+	list, err := net.ListenTCP("tcp", s.config.RPCAddr)
+	if err != nil || list == nil {
+		s.logger.Printf("[ERR] nomad: No TLS listener to reload")
+		return err
+	}
+
+	s.rpcListener = list
+
+	// reinitialize the cancel context
+	ctx, cancel := context.WithCancel(context.Background())
+	s.rpcCancel = cancel
+	s.rpcListenerLock.Unlock()
+
+	go s.listen(ctx)
+
+	s.raftLayerLock.Lock()
+	s.raftLayer.Close()
+	wrapper := tlsutil.RegionSpecificWrapper(s.config.Region, tlsWrap)
+	s.raftLayer = NewRaftLayer(s.rpcAdvertise, wrapper)
+	s.raftLayerLock.Unlock()
+
+	// re-initialize the network transport with a re-initialized stream layer
+	trans := raft.NewNetworkTransport(s.raftLayer, 3, s.config.RaftTimeout,
+		s.config.LogOutput)
+	s.raftTransport = trans
+
+	s.logger.Printf("[INFO] nomad: finished reloading server connections")
```

Comment:
Maybe demote this to DEBUG as in the common case everything should just work and operators should only need to know if an error occurs (which are logged separately).